### PR TITLE
Replace flat buttons with flat-styled buttons with a visible pressed state

### PR DESCRIPTION
**Pull request by YuriSizov**
`ToolButton` is back, baby!.. Wait, no, that sounds wrong. But we do have a usability problem which `ToolButton` could've solved.

### The problem

We use flat buttons throughout the editor UI, and in a couple of `Control` nodes (`GraphEdit` and `FileDialog`). The problem with flat buttons is that they by design don't render any styleboxes except for the focus outline. So no visible pressed state for them, except for slightly tinted icon and/or text. In the editor, by default, it's a light blue color, and in the default theme it's a slightly different shade of white (which makes the pressed state completely invisible in the `GraphEdit` toolbar).

image
_Which of these buttons is toggled on and which are toggled off? This is not a trick question!_

Within the editor users have expressed trouble noticing when some tools are enabled. This can depend a lot on the exact theme and theme colors used, but the default look definitely has this issue where lightly colored blue text or blue icon is getting lost on top of a bluish background.

https://youtu.be/nu6g942bAIU?t=846

### The solution

One way to solve this problem is to make flat buttons render the pressed state as well, but that would break compatibility and expectation compared to what we have right now.

Thankfully, we have theme type variations. So I propose we use them here and introduce the `FlatButton` type variation (and also `FlatMenuButton` for `MenuButton`), which would have empty styles for everything except for the pressed state. This would make them look like exactly like flat buttons, but with a proper style for when these buttons are pressed.

https://github.com/godotengine/godot/assets/11782833/5a660532-e4b1-4c4a-a4bc-8abcab571309

With this PR these variations are present in the default theme and used in `GraphEdit` and `FileDialog` as well:

godot windows editor dev x86_64_2023-09-19_17-58-29

_It's going to look even better with https://github.com/godotengine/godot/pull/81582._

### The implementation

I went through all toolbars and toolbar-alikes in the editor that use flat buttons and updated them to use this approach. Here are a few examples:

<details>
<summary>Videos and pictures!</summary>

https://github.com/godotengine/godot/assets/11782833/bea7a697-ba62-46fd-85e4-ec88d3c7ee70

https://github.com/godotengine/godot/assets/11782833/62395a3c-f1e7-4fe2-ab34-c2307e6f4105

godot windows editor dev x86_64_2023-09-19_17-58-56
godot windows editor dev x86_64_2023-09-19_18-00-46
godot windows editor dev x86_64_2023-09-19_18-01-13
</details>

I went mostly by code, and replaced flat buttons based on my memory of the tool, but I didn't check everything. I skipped some places too. So if you think I was too conservative or the other way around with the changes, do let me know and I will remove/add whatever is needed. Adjustments can also be done in a follow-up, of course.

----

Of course, the exact styling may require some adjustments. For example, it doesn't look too dissimilar from a non-flat button right now. But I think it's an overall usability improvement.

godot windows editor dev x86_64_2023-09-19_18-01-35

### CODE REVIEW

**Comment by Zireael07** [comment]
As a glasses user, THANK YOU <3

**Comment by MewPurPur** [comment]
Does light theme look good?

**Comment by KoBeWi** [comment]
Seems barely visible in light theme
image
Although the accent color is more visible.

File: scene/theme/default_theme.cpp [hunk 1]
```diff
@@ -360,6 +360,27 @@ void fill_default_theme(Ref<Theme> &theme, const Ref<Font> &default_font, const
...
 	theme->set_constant("outline_size", "CheckButton", 0);
 
+	// Button variations
+
+	theme->set_type_variation("FlatButton", "Button");
+	theme->set_type_variation("FlatMenuButton", "MenuButton");
+
+	Ref<StyleBoxFlat> flat_button_normal = button_normal->duplicate();
+	flat_button_normal->set_bg_color(Color(0, 0, 0, 0));
+	flat_button_normal->set_border_color(Color(0, 0, 0, 0));
```
Comment (KoBeWi) [inline scene/theme/default_theme.cpp:370]: You could make StyleboxEmpty and copy margins/border size from the normal stylebox.

Comment (YuriSizov) [reply scene/theme/default_theme.cpp:370]: Yes, I decided to do it this weird way because it would result in less upkeep if changes happen. Not that our theme files are well organized 🙁

File: editor/editor_themes.cpp [hunk 2]
```diff
@@ -1020,6 +1021,22 @@ Ref<Theme> create_editor_theme(const Ref<Theme> p_theme) {
 	theme->set_constant("h_separation", "Button", 4 * EDSCALE);
 	theme->set_constant("outline_size", "Button", 0);
 
+	Ref<StyleBoxFlat> style_flat_button = style_widget->duplicate();
+	style_flat_button->set_bg_color(Color(0, 0, 0, 0));
+	style_flat_button->set_border_color(Color(0, 0, 0, 0));
```
Comment (KoBeWi) [inline editor/editor_themes.cpp:1026]: Same here.

File: scene/theme/default_theme.cpp (hunk 1 above)
Comment (KoBeWi) [reply scene/theme/default_theme.cpp:370]: I mean, you can just do something like
```C++
for (int i = 0; i < 4; i++) {
	empty_box->set_content_margin(i, button_normal->get_margin(i) + button_normal->get_border_width(i))
}
```

**Comment by Calinou** [comment]
Could a translucent accent color be used for pressed flat buttons? This would make them easier to distinguish from other UI elements such as dropdowns.

**Comment by YuriSizov** [comment]
> Could a translucent accent color be used for pressed flat buttons? This would make them easier to distinguish from other UI elements such as dropdowns.

Like this?

godot windows editor dev x86_64_2023-09-19_22-33-20
godot windows editor dev x86_64_2023-09-19_22-33-46

----

godot windows editor dev x86_64_2023-09-19_22-33-06
godot windows editor dev x86_64_2023-09-19_22-35-05

File: scene/theme/default_theme.cpp (hunk 1 above)
Comment (YuriSizov) [reply scene/theme/default_theme.cpp]: Sure, I can do that.

File: editor/editor_themes.cpp [hunk 3]
```diff
@@ -949,7 +949,7 @@ Ref<Theme> create_editor_theme(const Ref<Theme> p_theme) {
 	toolbar_stylebox->set_border_color(accent_color);
 	toolbar_stylebox->set_border_width(SIDE_BOTTOM, Math::round(2 * EDSCALE));
 	toolbar_stylebox->set_content_margin(SIDE_BOTTOM, 0);
-	toolbar_stylebox->set_expand_margin_all(2 * EDSCALE);
+	toolbar_stylebox->set_expand_margin_individual(4 * EDSCALE, 2 * EDSCALE, 4 * EDSCALE, 4 * EDSCALE);
```
Comment (KoBeWi) [inline editor/editor_themes.cpp:952]: Why this change?

Comment (YuriSizov) [reply editor/editor_themes.cpp:952]: That stylebox is actually clipping buttons, but it's not noticeable if you don't have styles rendered on them. And the sides were too close with the pressed style on, so I extended them a bit.

Comment (KoBeWi) [reply editor/editor_themes.cpp:952]: image
Without this change, the side margins match the top margin (bottom does look bad though).

**Comment by Calinou** [comment]
> Like this?

Looks good! Maybe the blue accent could be a bit more prominent.

**Review by KoBeWi — APPROVED** [review]
The new style and implementation looks good.
I didn't check whether we should apply it to more buttons.

**Comment by YuriSizov** [comment]
> Maybe the blue accent could be a bit more prominent, as it's quite subtle right now and can be hard to see (especially under sunlight).

I get why you would want this, but the trouble is the content is also colored with the accent color. The closer we get to it, the less recognizable the content is going to be. There may be another way to style it entirely, without coloring the icon and the text with the accent color, perhaps.

We could probably bikeshed and experiment more, but it's also something that we can tune afterwards, I think.

**Comment by MewPurPur** [comment]
On the contrary, I think we're at a good place between it being obvious and it not sticking out too much.
